Cannot play Bluray on latest release on debian 12

[debug_20.5.0.log](https://github.com/user-attachments/files/18751075/debug_20.5.0.log)

## Bug report
### Describe the bug
Using an external Bluray drive, valid bluray that plays using VLC do not play using xbmc


## Expected Behavior
Insert disc in drive, go to "Drive" menu and then start playing the disc


## Actual Behavior
The disc doesn't play.
Even its title is not displayed



## Possible Fix
Using debug log, is appears that xbmc tries to open /dev/cdrom/<path> instead of using disc mount point.


### To Reproduce
Steps to reproduce the behavior:
1. Start kodi
2. Insert disc 
3. Go to disc menu 


### Debuglog
2025-02-13 12:59:49.562 T:42236 debug : CBlurayCallback::Logger - bluray.c:1463: libbluray version 1.3.4
2025-02-13 12:59:49.562 T:42236 debug : CBlurayCallback::Logger - bluray.c:1488: BLURAY initialized!
2025-02-13 12:59:49.562 T:42236 debug : CBlurayCallback::Logger - register.c:420: bd_psr_write(): PSR18 (MENU_LANG) 0xffffff -> 0x656e67
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback::Logger - disc.c:333: failed opening UDF image (null)
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Error opening file! (/dev/cdrom/AACS/Unit_Key_RO.inf)
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback::Logger - aacs.c:83: AACS/Unit_Key_RO.inf not found. No AACS protection.
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Error opening file! (/dev/cdrom/BDMV/index.bdmv)
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Error opening file! (/dev/cdrom/BDMV/INDEX.BDM)
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback::Logger - disc.c:437: error opening file BDMV/index.bdmv
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Error opening file! (/dev/cdrom/BDMV/BACKUP/index.bdmv)
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback::Logger - disc.c:437: error opening file BDMV/BACKUP/index.bdmv
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Opening dir /dev/cdrom/BDMV/PLAYLIST
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Error opening dir! (/dev/cdrom/BDMV/PLAYLIST)
2025-02-13 12:59:49.563 T:42236 error : CBlurayDirectory::InitializeBluray - failed to open /dev/cdrom
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback::Logger - bluray.c:1610: BLURAY destroyed!

[debug.log](https://github.com/user-attachments/files/18783439/debug.log)


### Your Environment
Used Operating system:
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- Put your text below this line. Checkboxes can easily be ticked once issue is created -->
 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [x] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: sgo@pi5:~ $ raspbian 12.9
 - Linux pi5 6.6.62+rpt-rpi-2712 #1 SMP PREEMPT Debian 1:6.6.62-1+rpt1 (2024-11-25) aarch64 GNU/Linux
 - Kodi version: Kodi Media Center 22.0-ALPHA1 (21.90.700) Git:20250210-0bf4dd5f0c




<!--- End of this issue -->
*n*ote: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.

The same issue is reproduced on 

- Kodi version: Kodi Media Center 22.0-ALPHA1 (21.90.700) Git:20250210-0bf4dd5f0c

Debuglog

2025-02-13 12:59:49.562 T:42236 debug : CBlurayCallback::Logger - bluray.c:1463: libbluray version 1.3.4
2025-02-13 12:59:49.562 T:42236 debug : CBlurayCallback::Logger - bluray.c:1488: BLURAY initialized!
2025-02-13 12:59:49.562 T:42236 debug : CBlurayCallback::Logger - register.c:420: bd_psr_write(): PSR18 (MENU_LANG) 0xffffff -> 0x656e67
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback::Logger - disc.c:333: failed opening UDF image (null)
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Error opening file! (/dev/cdrom/AACS/Unit_Key_RO.inf)
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback::Logger - aacs.c:83: AACS/Unit_Key_RO.inf not found. No AACS protection.
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Error opening file! (/dev/cdrom/BDMV/index.bdmv)
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Error opening file! (/dev/cdrom/BDMV/INDEX.BDM)
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback::Logger - disc.c:437: error opening file BDMV/index.bdmv
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Error opening file! (/dev/cdrom/BDMV/BACKUP/index.bdmv)
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback::Logger - disc.c:437: error opening file BDMV/BACKUP/index.bdmv
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Opening dir /dev/cdrom/BDMV/PLAYLIST
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback - Error opening dir! (/dev/cdrom/BDMV/PLAYLIST)
2025-02-13 12:59:49.563 T:42236 error : CBlurayDirectory::InitializeBluray - failed to open /dev/cdrom
2025-02-13 12:59:49.563 T:42236 debug : CBlurayCallback::Logger - bluray.c:1610: BLURAY destroyed!

[https://github.com/user-attachments/files/18783439/debug.log](url)

--- a/cmake/scripts/linux/ArchSetup.cmake
+++ b/cmake/scripts/linux/ArchSetup.cmake
@@ -156,6 +156,11 @@ if(HAVE_MKOSTEMP)
   list(APPEND ARCH_DEFINES "-DHAVE_MKOSTEMP=1")
 endif()
 
+check_symbol_exists(getmntent_r "mntent.h" HAVE_GETMNTENT_R)
+if(HAVE_GETMNTENT_R)
+  list(APPEND ARCH_DEFINES "-DHAVE_GETMNTENT_R=1")
+endif()
+
 set(CMAKE_REQUIRED_DEFINITIONS "-D_GNU_SOURCE")
 check_symbol_exists("memfd_create" "sys/mman.h" HAVE_LINUX_MEMFD)
 set(CMAKE_REQUIRED_DEFINITIONS "")

--- a/xbmc/Autorun.cpp
+++ b/xbmc/Autorun.cpp
@@ -116,13 +116,12 @@ bool CAutorun::PlayDisc(const std::string& path, bool bypassSettings, bool start
   std::string mediaPath;
 
   CCdInfo* pInfo = CServiceBroker::GetMediaManager().GetCdInfo(path);
-  if (pInfo == NULL)
-    return false;
 
-  if (pInfo->IsAudio(1))
+  if (pInfo && pInfo->IsAudio(1))
     mediaPath = "cdda://local/";
 
-  if (mediaPath.empty() && (pInfo->IsISOUDF(1) || pInfo->IsISOHFS(1) || pInfo->IsIso9660(1) || pInfo->IsIso9660Interactive(1)))
+  if (mediaPath.empty() && (pInfo && (pInfo->IsISOUDF(1) || pInfo->IsISOHFS(1) ||
+                                      pInfo->IsIso9660(1) || pInfo->IsIso9660Interactive(1))))
     mediaPath = "iso9660://";
 
   if (mediaPath.empty())

--- a/xbmc/filesystem/BlurayDirectory.cpp
+++ b/xbmc/filesystem/BlurayDirectory.cpp
@@ -14,6 +14,7 @@
 #include "URL.h"
 #include "filesystem/BlurayCallback.h"
 #include "filesystem/Directory.h"
+#include "filesystem/DirectoryFactory.h"
 #include "guilib/LocalizeStrings.h"
 #include "utils/LangCodeExpander.h"
 #include "utils/RegExp.h"
@@ -284,7 +285,13 @@ bool CBlurayDirectory::InitializeBluray(const std::string &root)
   g_LangCodeExpander.ConvertToISO6392T(g_langInfo.GetDVDMenuLanguage(), langCode);
   bd_set_player_setting_str(m_bd, BLURAY_PLAYER_SETTING_MENU_LANG, langCode.c_str());
 
-  if (!bd_open_files(m_bd, const_cast<std::string*>(&root), CBlurayCallback::dir_open, CBlurayCallback::file_open))
+  m_realPath = root;
+
+  auto fileHandler = CDirectoryFactory::Create(CURL{root});
+  if (fileHandler)
+    m_realPath = fileHandler->ResolveMountPoint(root);
+
+  if (!bd_open_files(m_bd, &m_realPath, CBlurayCallback::dir_open, CBlurayCallback::file_open))
   {
     CLog::Log(LOGERROR, "CBlurayDirectory::InitializeBluray - failed to open {}",
               CURL::GetRedacted(root));

--- a/xbmc/filesystem/BlurayDirectory.h
+++ b/xbmc/filesystem/BlurayDirectory.h
@@ -51,6 +51,7 @@ private:
   CURL          m_url;
   BLURAY*       m_bd = nullptr;
   bool          m_blurayInitialized = false;
+  std::string m_realPath;
 };
 
 }

--- a/xbmc/filesystem/IDirectory.h
+++ b/xbmc/filesystem/IDirectory.h
@@ -104,6 +104,13 @@ public:
   virtual bool Remove(const CURL& url) { return false; }
 
   /*!
+  \brief Provided a path, attempts to resolve to a mount point
+  \param path Path to resolve
+  \return Returns the mountpoint if found, else the provided path
+  */
+  virtual std::string ResolveMountPoint(const std::string& path) const { return path; }
+
+  /*!
   \brief Recursively removes the directory
   \param url Directory to remove.
   \return Returns \e false if not successful

--- a/xbmc/platform/posix/filesystem/PosixDirectory.cpp
+++ b/xbmc/platform/posix/filesystem/PosixDirectory.cpp
@@ -16,8 +16,12 @@
 #include "utils/StringUtils.h"
 #include "utils/URIUtils.h"
 #include "utils/XTimeUtils.h"
+#include "utils/log.h"
 
 #include <dirent.h>
+#ifdef HAVE_GETMNTENT_R
+#include <mntent.h>
+#endif
 #include <sys/stat.h>
 
 using namespace XFILE;
@@ -205,4 +209,52 @@ bool CPosixDirectory::Exists(const CURL& url)
   if (stat(path.c_str(), &buffer) != 0)
     return false;
   return S_ISDIR(buffer.st_mode) ? true : false;
+}
+
+std::string CPosixDirectory::ResolveMountPoint(const std::string& file) const
+{
+#ifdef HAVE_GETMNTENT_R
+  // Check if file is a block device, if yes we need the mount point
+  struct stat st;
+
+  if (stat(file.c_str(), &st) == 0 && S_ISBLK(st.st_mode))
+  {
+    // the path may be a symlink (e.g. /dev/dvd -> /dev/sr0), so make sure
+    // we look up the real device
+    char pathBuffer[MAX_PATH];
+    char* bdDevice = realpath(file.c_str(), pathBuffer);
+
+    if (bdDevice == nullptr)
+    {
+      CLog::LogF(LOGERROR, "realpath on '{}' failed with {}", file, errno);
+      return file;
+    }
+
+    struct FileDeleter
+    {
+      void operator()(FILE* f) { endmntent(f); }
+    };
+    std::unique_ptr<FILE, FileDeleter> mtab(setmntent("/proc/self/mounts", "r"));
+
+    if (mtab)
+    {
+      mntent *m, mbuf;
+      char buf[8192];
+      std::string_view bdDeviceSv = bdDevice;
+
+      while ((m = getmntent_r(mtab.get(), &mbuf, buf, sizeof(buf))) != nullptr)
+      {
+        if (bdDeviceSv == m->mnt_fsname)
+        {
+          return m->mnt_dir;
+        }
+      }
+
+      CLog::LogF(LOGINFO, "No mount point for device '{}' found", bdDeviceSv);
+    }
+    else
+      CLog::LogF(LOGWARNING, "Failed to open mount table");
+  }
+#endif
+  return file;
 }

--- a/xbmc/platform/posix/filesystem/PosixDirectory.h
+++ b/xbmc/platform/posix/filesystem/PosixDirectory.h
@@ -23,6 +23,8 @@ public:
   bool Exists(const CURL& url) override;
   bool Remove(const CURL& url) override;
   bool RemoveRecursive(const CURL& url) override;
+  std::string ResolveMountPoint(const std::string& file) const override;
+
 private:
   bool Create(const std::string& path);
 };

